docs/content/guides/dagster/transitioning-data-pipelines-from-development-to-production.mdx

```diff
@@ -0,0 +1,456 @@
+                "user": "system@company.com",
+                "password": {"env": "SYSTEM_SNOWFLAKE_PASSWORD"},
+                "database": "STAGING",
+                "schema": "HACKER_NEWS",
+            }
+        ),
+    },
+}
+```
+
+---
+
+## Optional: Testing
+
+You may have noticed a missing step in the development workflow presented in this guide — unit tests! While the main purpose of the guide is to help you transition your code from local development to a production deployment, unit testing is still an important part of the development cycle. In this section, we'll explore a pattern you may find useful when writing your own unit tests.
+
+When we write unit tests for the `items` asset, we could make more precise assertions if we knew exactly what data we'd receive from Hacker News. If we refactor our interactions with the Hacker News API as a resource, we can leverage Dagster's resource system to provide a stub resource in our unit tests.
+
+Before we get into implementation, let's go over some best practices:
+
+### When to use resources
+
+In many cases, it is often most convenient to interact with an external service directly in assets or ops, rather than write a resource. We recommend refactoring code to use resources in the following cases:
+
+- Multiple assets or ops need to interact with the service in a consistent way
+- Different implementations of a service need to be used in certain scenarios (ie. a staging environment, or unit tests)
+
+### When to use stub resources
+
+Determining when it makes sense to stub a resource for a unit test can be a topic of much debate. There are certainly some resources where it would be too complicated to write and maintain a stub. For example, it would be difficult to mock a database like Snowflake with a lightweight database since the SQL syntax may vary. In general, if a resource is relatively simple, writing a stub can be really helpful for unit testing the assets and ops that use the resource.
+
+We'll start by writing the "real" Hacker News API Client:
+
+```python file=/guides/dagster/development_to_production/resources/resources_v1.py startafter=start_resource endbefore=end_resource
+# resources.py
+from typing import Any, Dict, Optional
+
+import requests
+
+from dagster import resource
+
+
+class HNAPIClient:
+    """
+    Hacker News client that fetches live data
+    """
+
+    def fetch_item_by_id(self, item_id: int) -> Optional[Dict[str, Any]]:
+        """Fetches a single item from the Hacker News API by item id."""
+
+        item_url = f"https://hacker-news.firebaseio.com/v0/item/{item_id}.json"
+        item = requests.get(item_url, timeout=5).json()
+        return item
+
+    def fetch_max_item_id(self) -> int:
+        return requests.get(
+            "https://hacker-news.firebaseio.com/v0/maxitem.json", timeout=5
+        ).json()
+
+
+@resource
+def hn_api_client():
+    return HNAPIClient()
+```
+
+We'll also need to update the `items` asset to use this resource:
```

Comment:
I think it would be helpful here to include a line that readers will also have to add the resource to the `resource_defs` field within their repository